### PR TITLE
Adopt the Whitaker Dylint suite in the lint gate and CI

### PR DESCRIPTION
**Pull request by leynos**
## Summary

This branch adopts the [Whitaker](https://github.com/leynos/whitaker) Dylint suite as part of the estate-wide lint rollout, following the pattern established in leynos/wireframe. The suite now runs clean over `--all-targets --all-features`, is enforced by `make lint`, and runs in CI on the stable matrix leg. Roughly 240 findings across four waves (build script → library → library tests → integration test crates) were fixed rather than suppressed, with two documented crate-level exclusions for deliberately ambient filesystem access.

The adoption also surfaced a Whitaker defect — `no_std_fs_operations` ignores in-source `allow`/`expect` attributes — filed as leynos/whitaker#270. Crate-granularity exclusion via `dylint.toml` is therefore the only working escape hatch, which shaped the design below.

## Review walkthrough

- Start with [dylint.toml](https://github.com/leynos/netsuke/blob/1900485/dylint.toml) for the lint policy: each `no_std_fs_operations` exclusion carries its rationale (build script, the two boundary crates, and the fixture-staging integration test crates).
- Then review [ambient_fs/src/lib.rs](https://github.com/leynos/netsuke/blob/1900485/ambient_fs/src/lib.rs), a new internal leaf crate confining the probes the `which` resolver genuinely cannot express through `cap_std` handles (a capability handle refuses to follow symlinks that leave its directory, which would break layouts such as `/usr/bin/cc -> /etc/alternatives/cc`). Its scope and reuse policy are documented in the crate docs. [test_support/src/fs.rs](https://github.com/leynos/netsuke/blob/1900485/test_support/src/fs.rs) is the analogous home for ambient test-fixture I/O, mirroring Whitaker's own `whitaker_common` exclusion.
- [src/manifest/mod.rs](https://github.com/leynos/netsuke/blob/1900485/src/manifest/mod.rs) shows the main behavioural change in production code: `open_manifest_workspace` opens the workspace's `cap_std` handle first and reads the manifest through it, instead of an ambient `fs::read_to_string`.
- [src/stdlib/config.rs](https://github.com/leynos/netsuke/blob/1900485/src/stdlib/config.rs) replaces the panicking `Default for StdlibConfig` with a fallible `StdlibConfig::from_current_dir()`; [src/stdlib/command/config.rs](https://github.com/leynos/netsuke/blob/1900485/src/stdlib/command/config.rs) switches the pipe limit tracker to saturating arithmetic so overflow trips the limit check rather than panicking.
- [src/stdlib/network/mod.rs](https://github.com/leynos/netsuke/blob/1900485/src/stdlib/network/mod.rs) and [src/stdlib/network/cache.rs](https://github.com/leynos/netsuke/blob/1900485/src/stdlib/network/cache.rs) show the `bumpy_road_function` refactors and the resulting module split; [src/ir/cycle.rs](https://github.com/leynos/netsuke/blob/1900485/src/ir/cycle.rs) shows the largest `module_max_lines` split (682 → 377 lines, helpers in [src/ir/cycle_support.rs](https://github.com/leynos/netsuke/blob/1900485/src/ir/cycle_support.rs)).
- The bulk of the diff is mechanical test-code compliance: fixtures return `Result` and tests consume them with `?` (assertions become `anyhow::ensure!` to satisfy `clippy::panic_in_result_fn`), fixture I/O routes through `test_support::fs`, and inline test modules gain `//!` docs.
- Finish with [Makefile](https://github.com/leynos/netsuke/blob/1900485/Makefile) and [.github/workflows/ci.yml](https://github.com/leynos/netsuke/blob/1900485/.github/workflows/ci.yml) for the gate wiring: `make lint` runs rustdoc, Clippy, and Whitaker; CI installs `whitaker-installer` 0.2.5 via a cached `cargo binstall` on the stable leg only, while the MSRV and nightly legs run the new `make lint-clippy` subset (Whitaker lints under its own pinned toolchain, so repeating it per leg duplicates identical work). Documentation updates live in [docs/developers-guide.md](https://github.com/leynos/netsuke/blob/1900485/docs/developers-guide.md), the vendored [docs/whitaker-users-guide.md](https://github.com/leynos/netsuke/blob/1900485/docs/whitaker-users-guide.md), and [AGENTS.md](https://github.com/leynos/netsuke/blob/1900485/AGENTS.md).

## Validation

- `make check-fmt`: pass
- `make lint` (rustdoc + Clippy + `whitaker --all -- --all-targets --all-features`, warnings denied): pass, zero Whitaker findings
- `make typecheck`: pass
- `make test`: pass (all targets, all features)
- `make test-workflow-contracts`: 6 passed
- `make markdownlint`: 70 files, 0 errors
- `make nixie`: all Mermaid diagrams valid
- `mbake validate Makefile`: valid

## Notes

- In-source suppression (`cfg_attr(dylint_lib = …, expect(…))`) is non-functional in Whitaker v0.2.5 (leynos/whitaker#270); the `cfg(dylint_lib, values(any()))` check-cfg allowlist entry is retained so attribute-based suppression can replace crate exclusions once the lint honours it.
- New integration test crates that touch `std::fs` will fail the lint until deliberately added to `dylint.toml`, making that policy decision visible in review.
- Adding the `ambient_fs` crate required a matching `.github/dependabot.yml` entry (enforced by `dependabot_config_tests`).

## References

- [Lody session](https://lody.ai/leynos/sessions/41883f0a-5fa9-4129-8f4b-edb5613ccf61)

### CODE REVIEW

**Comment by leynos** [comment]
@coderabbitai Please suggest a fix for this issue and supply a prompt for an AI coding agent to enable it to apply the fix. Include the file and symbol names indicated in the issue at the head of your response. Ensure that this is validated against the *current* version of the codegraph.

If further refinement to address this finding would be deleterious, please supply a clear explanatory one to two paragraph markdown message I can paste into the CodeScene web ui's diagnostic suppression function so this diagnostic can be silenced.

[src/stdlib/network/tests.rs](https://github.com/leynos/netsuke/pull/410/files/1900485435b521ac4863a51224d4c29ce831f071#diff-21fc21b945a1fa115fb32e8612c4618a2c54484268c642e5a964e646d244b8d3)

Comment on file

~~~rust
        Arc,
        atomic::{AtomicBool, Ordering},
    },
use std::sync::{
~~~

❌ New issue: [Code Duplication](https://codescene.io/projects/69281/delta?repo-id=1160466&review-id=410&biomarker=Code+Duplication&filename=src%2Fstdlib%2Fnetwork%2Ftests.rs&method=)
The module contains 2 functions with similar structure: fetch_rejects_disallowed_scheme,fetch_rejects_not_allowlisted_host

**Comment by leynos** [comment]
@coderabbitai Please suggest a fix for this issue and supply a prompt for an AI coding agent to enable it to apply the fix. Include the file and symbol names indicated in the issue at the head of your response. Ensure that this is validated against the *current* version of the codegraph.

If further refinement to address this finding would be deleterious, please supply a clear explanatory one to two paragraph markdown message I can paste into the CodeScene web ui's diagnostic suppression function so this diagnostic can be silenced.

[tests/cli_tests/config_discovery_scopes.rs](https://github.com/leynos/netsuke/pull/410/files/1900485435b521ac4863a51224d4c29ce831f071#diff-b786ea54347cf6dbaa26ae590d141d8f36f037538f1b3ac8508924c08dcfb934)

Comment on lines +252 to +295

~~~rust
fn project_config_takes_precedence_over_user_config() -> Result<()> {
    let _env_lock = EnvLock::acquire();
    let cwd_guard = CwdGuard::acquire().context("capture current working directory")?;

    let temp_project = tempdir().context("create temporary project directory")?;
    let temp_appdata = tempdir().context("create temporary APPDATA directory")?;

    // Create sandboxed Windows user-scope config at %APPDATA%\netsuke\config.toml
    let netsuke_config_dir = temp_appdata.path().join("netsuke");
    fs::create_dir_all(&netsuke_config_dir).context("create netsuke config directory")?;
    fs::write(
        netsuke_config_dir.join("config.toml"),
        PRECEDENCE_USER_CONFIG_CONTENT,
    )
    .context("write user config.toml in APPDATA")?;

    // Project config: overrides theme; does NOT set colour_policy.
    fs::write(
        temp_project.path().join(".netsuke.toml"),
        PRECEDENCE_PROJECT_CONFIG_CONTENT,
    )
    .context("write project .netsuke.toml")?;

    let _appdata_guard = EnvVarGuard::set("APPDATA", temp_appdata.path().as_os_str());
    let _localappdata_guard = EnvVarGuard::remove("LOCALAPPDATA");
    let _config_guard = EnvVarGuard::remove("NETSUKE_CONFIG");
    let _config_path_guard = EnvVarGuard::remove("NETSUKE_CONFIG_PATH");
    let _theme_guard = EnvVarGuard::remove("NETSUKE_THEME");
    let _jobs_guard = EnvVarGuard::remove("NETSUKE_JOBS");
    let _colour_guard = EnvVarGuard::remove("NETSUKE_COLOUR_POLICY");

    std::env::set_current_dir(&temp_project).context("change to project directory")?;

    let localizer = Arc::from(cli_localization::build_localizer(None));
    let (cli, matches) = netsuke::cli::parse_with_localizer_from(["netsuke"], &localizer)
        .context("parse CLI for precedence test")?;
    let merged = netsuke::cli::merge_with_config(&cli, &matches)
        .context("merge configs")?
        .with_default_command();

    let result = assert_project_precedence_applied(&merged);
    drop(cwd_guard);
    result
}
~~~

❌ New issue: [Code Duplication](https://codescene.io/projects/69281/delta?repo-id=1160466&review-id=410&biomarker=Code+Duplication&filename=tests%2Fcli_tests%2Fconfig_discovery_scopes.rs&method=)
The module contains 2 functions with similar structure: project_config_takes_precedence_over_user_config,project_config_takes_precedence_over_user_config

File: Cargo.toml [hunk 1]
```diff
@@ -53,6 +53,7 @@ tracing-subscriber = { version = "0.3", features = ["fmt"] }
 serde_json = { version = "1", features = ["preserve_order"] }
 serde_json_canonicalizer = "0.3"
 tempfile = "3.8.0"
+ambient_fs = { path = "ambient_fs" }
```
[the remark this answers is not in the document]
Comment (leynos) [reply Cargo.toml:63]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.

**Comment by pandalump** [comment]
@coderabbitai resume

**Comment by leynos** [comment]
@coderabbitai resume

File: tests/bdd/steps/manifest_command_helpers.rs [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+    let exe = assert_cmd::cargo::cargo_bin!("netsuke");
+    ensure!(
+        exe.is_file(),
+        "netsuke binary not found at {}",
+        exe.display()
+    );
+    Ok(exe.to_path_buf())
+}
+
+/// Build a netsuke command with a sanitised environment.
```
[the remark this answers is not in the document]
Comment (leynos) [reply tests/bdd/steps/manifest_command_helpers.rs]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.

File: tests/bdd/steps/manifest_command_helpers.rs [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+    store_run_result(
+        world,
+        RunResult {
+            stdout: String::from_utf8_lossy(&output.stdout).into_owned(),
+            stderr: String::from_utf8_lossy(&output.stderr).into_owned(),
+            success: output.status.success(),
+        },
+    );
+    Ok(())
+}
```
[the remark this answers is not in the document]
Comment (leynos) [reply tests/bdd/steps/manifest_command_helpers.rs:182]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.

File: src/stdlib/path/hash_utils.rs [hunk 4]
```diff
@@ -88,10 +88,10 @@ where
...
             break;
         }
-        let chunk = buffer.get(..read).unwrap_or_else(|| {
-            debug_assert!(false, "read beyond buffer capacity: {read} bytes");
-            buffer.as_slice()
-        });
+        // A well-behaved Read never reports more bytes than the buffer holds;
+        // clamp to the full buffer rather than panicking on a misbehaving
+        // implementation.
+        let chunk = buffer.get(..read).unwrap_or(&buffer);
```
[the remark this answers is not in the document]
Comment (leynos) [reply src/stdlib/path/hash_utils.rs]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.

File: src/stdlib/config.rs [hunk 5]
```diff
@@ -328,26 +328,32 @@ impl StdlibConfig {
...
         let path = Utf8PathBuf::from_path_buf(cwd)
-            .unwrap_or_else(|path| panic!("cwd contains non-UTF-8 components: {}", path.display()));
+            .map_err(|path| anyhow!("cwd contains non-UTF-8 components: {}", path.display()))?;
         Self::new(root)
-            .unwrap_or_else(|err| panic!("default fetch cache path should be valid: {err}"))
+            .context("default fetch cache path should be valid")?
             .with_workspace_root_path(path)
-            .unwrap_or_else(|err| panic!("workspace root must be absolute: {err}"))
+            .context("workspace root must be absolute")
     }
```
[the remark this answers is not in the document]
Comment (leynos) [reply src/stdlib/config.rs:365]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.

File: src/stdlib/config_tests.rs [hunk 6]
```diff
@@ -0,0 +1,199 @@
...
+}
+
+#[fixture]
+fn base_config(#[from(workspace)] workspace: Result<(Dir, Utf8PathBuf)>) -> Result<StdlibConfig> {
+    let (dir, path) = workspace?;
+    StdlibConfig::new(dir)
+        .context("construct stdlib config")?
+        .with_workspace_root_path(path)
+        .context("record workspace root")
+}
```
[the remark this answers is not in the document]
Comment (leynos) [reply src/stdlib/config_tests.rs:16]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.

File: src/stdlib/command/pipes.rs [hunk 7]
```diff
@@ -266,34 +272,38 @@ mod tests {
...
-        assert_eq!(disk, payload);
-        fs::remove_file(path.as_std_path()).expect("cleanup streamed file");
+        let disk = fs::read(path.as_std_path()).context("read streamed output")?;
+        ensure!(
+            disk == payload,
+            "streamed file contents did not match payload"
+        );
+        fs::remove_file(path.as_std_path()).context("cleanup streamed file")?;
+        Ok(())
     }
```
[the remark this answers is not in the document]
Comment (leynos) [reply src/stdlib/command/pipes.rs:279]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.

File: src/stdlib/command/pipes.rs [hunk 8]
```diff
@@ -199,25 +202,28 @@ fn persist_tempfile(tempfile: NamedTempFile) -> Result<Utf8PathBuf, CommandFailu
...
+            Dir::open_ambient_dir(&path, ambient_authority()).context("open temp workspace dir")?;
         let config = CommandConfig::new(
             DEFAULT_COMMAND_MAX_OUTPUT_BYTES,
             DEFAULT_COMMAND_MAX_STREAM_BYTES,
             Arc::new(dir),
             Some(Arc::new(path)),
         );
-        (temp, config)
+        Ok((temp, config))
     }
```
[the remark this answers is not in the document]
Comment (leynos) [reply src/stdlib/command/pipes.rs]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.

File: src/runner/process/file_io.rs [hunk 9]
```diff
@@ -187,16 +188,10 @@ mod tests {
...
-            .read_to_string(&mut written)
-            .context("re-read reopened temp file")?;
+        let reread = test_fs::read_to_string(file.path()).context("re-read temp file")?;
         ensure!(
-            written == content.as_str(),
-            "re-read file contents '{written}' did not match '{expected}'",
+            reread == content.as_str(),
+            "re-read file contents '{reread}' did not match '{expected}'",
             expected = content.as_str()
         );
```
[the remark this answers is not in the document]
Comment (leynos) [reply src/runner/process/file_io.rs]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.

File: .github/workflows/ci.yml [hunk 10]
```diff
@@ -42,8 +43,34 @@ jobs:
...
+        run: |
+          if ! command -v whitaker-installer >/dev/null 2>&1; then
+            if cargo binstall --version >/dev/null 2>&1; then
+              cargo binstall --no-confirm --locked "whitaker-installer@${WHITAKER_INSTALLER_VERSION}"
+            else
+              echo "cargo-binstall unavailable; building whitaker-installer from crates.io"
+              cargo install --locked whitaker-installer --version "${WHITAKER_INSTALLER_VERSION}"
+            fi
+          fi
+          whitaker-installer
```
[the remark this answers is not in the document]
Comment (leynos) [reply .github/workflows/ci.yml:65]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.

**Comment by leynos** [comment]
@coderabbitai Have the following failed checks now been resolved?

If further work is required, please provide an AI agent prompt for the remaining work to be done to address these failures.

Do not treat warnings as optional or aspirational. Where a change is out of scope for this PR, propose a GitHub issue unless one exists already. (Treat o11y, documentation and validation coverage as in scope).

### ❌ Failed checks (2 errors, 1 warning)

| Check name | Status | Explanation | Resolution |
| :-: | :-- | :-- | :-- |
| Testing (Overall) | ❌ Error | The `canonicalize_resolves_to_absolute_path` test is vacuous: a no-op that returns the already-absolute tempdir path would still pass. | Use a relative path or symlink target so only real canonicalisation passes, and add a behavioural oracle for `sync_file` or explicitly justify a smoke test. |
| Rust Compiler Lint Integrity | ❌ Error | tests/bdd/types.rs keeps a blanket `#![expect(clippy::allow_attributes)]` plus `#[allow(dead_code)]` on macro-generated accessors, hiding real unused helper methods. | Remove the blanket suppression; delete unused accessors or gate only the specific, issue-linked items with narrow `#[expect(dead_code)]` where a real tracked use exists. |
| Observability | ⚠️ Warning | FAIL: The new fetch/cache and current-dir config paths return contextual errors but add no tracing/metrics, and the fetch path has no hit/miss visibility beyond one cleanup warning. | Add tracing or structured logging at fetch/cache hit-miss, cache open/read/write, and manifest/current-dir opening boundaries; include stable fields like URL/key, workspace path, and error category. Add metrics if cache behaviour must be... |

**Comment by leynos** [comment]
@coderabbitai Have the following now been resolved?

#### ambient_fs/src/lib.rs (1)

`1-127`: _📐 Maintainability & Code Quality_ | _🔵 Trivial_ | _⚡ Quick win_
**Add unit tests for this new crate.**
Ship this crate with its own `#[cfg(test)]` module rather than leaning on downstream `which` resolver tests for indirect coverage. Pin down `is_executable_file`'s execute-bit logic on Unix (regular file with/without execute bits, missing path, directory) and the non-Unix fallback, plus a happy-path check for `canonicalize` and `sync_file` using `tempfile`. This is a foundational shared utility now excluded from the `no_std_fs_operations` lint, so its contract deserves direct verification rather than relying on transitive exercise through `which`.
As per coding guidelines, "New functionality or changes in behaviour are fully validated by relevant unit tests and behavioural tests" and "Ensure that new features are validated with unit tests using `rstest`... Cover happy paths, unhappy paths, and relevant edge cases."

##### 🧪 Suggested test skeleton

```rust
#[cfg(test)]
mod tests {
    use super::*;
    use rstest::rstest;
    use tempfile::tempdir;

    #[rstest]
    fn canonicalize_resolves_relative_dot() {
        let canonical = canonicalize(Utf8Path::new(".")).expect("cwd resolves");
        assert!(canonical.is_absolute());
    }

    #[cfg(unix)]
    #[rstest]
    fn is_executable_file_detects_execute_bit() {
        use std::os::unix::fs::PermissionsExt;
        let dir = tempdir().expect("tempdir");
        let path = Utf8Path::from_path(dir.path().join("script")).expect("utf8 path");
        std::fs::write(&path, b"#!/bin/sh\n").expect("write file");
        std::fs::set_permissions(&path, std::fs::Permissions::from_mode(0o755))
            .expect("set perms");
        assert!(is_executable_file(&path).expect("metadata reads"));
    }
}
```

##### 🤖 Detailed instructions

Verify each finding against current code. Fix only still-valid issues, skip the
rest with a brief reason, keep changes minimal, and validate.

In `@ambient_fs/src/lib.rs` around lines 1 - 127, Add a dedicated #[cfg(test)]
module in ambient_fs and cover the public helpers directly instead of relying on
downstream tests. Use rstest/tempfile to verify is_executable_file for Unix
execute-bit behavior (regular file with and without execute bits, missing path,
and directory) plus the non-Unix regular-file fallback, and add happy-path tests
for canonicalize and sync_file. Keep the tests local to the crate and reference
the existing functions is_executable_file, canonicalize, and sync_file so the
contract is validated at the source.

_Source: Coding guidelines_

**Comment by leynos** [comment]
@coderabbitai resume

**Comment by pandalump** [comment]
@coderabbitai review

File: src/manifest/mod.rs [hunk 11]
```diff
@@ -250,32 +256,57 @@ fn resolve_absolute_workspace_root(utf8_parent: &Utf8Path) -> Result<Utf8PathBuf
...
-        .with_network_policy(policy))
+                .with_arg("workspace", root.as_str())
+                .with_arg("manifest", &manifest_file)
+        })?;
+    Ok(ManifestWorkspace {
+        dir,
+        root,
+        manifest_file,
+    })
 }
```
[the remark this answers is not in the document]
Comment (leynos) [reply src/manifest/mod.rs:317]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.

**Comment by leynos** [comment]
@coderabbitai Have the following now been resolved?

#### test_support/src/lib.rs (1)

`121-134`: _📐 Maintainability & Code Quality_ | _🟠 Major_ | _⚡ Quick win_
**Use the new `fs::set_mode` helper instead of re-deriving it here.**
`fake_ninja`'s Unix branch still hand-rolls metadata → permissions → `set_mode(0o755)` → `set_permissions`, but `test_support::fs::set_mode` (added in this same PR, see `test_support/src/fs.rs`) does precisely that. Wire this call through the new helper and drop the now-redundant `PermissionsExt` import.

##### ♻️ Proposed fix

```diff
-        use std::os::unix::fs::PermissionsExt;
-        let mut perms = std::fs::metadata(&path)
-            .with_context(|| format!("fake_ninja: read metadata {}", path.display()))?
-            .permissions();
-        perms.set_mode(0o755);
-        std::fs::set_permissions(&path, perms)
-            .with_context(|| format!("fake_ninja: set permissions {}", path.display()))?;
+        fs::set_mode(&path, 0o755)
+            .with_context(|| format!("fake_ninja: set permissions {}", path.display()))?;
```

##### 🤖 Detailed instructions

Verify each finding against current code. Fix only still-valid issues, skip the
rest with a brief reason, keep changes minimal, and validate.

In `@test_support/src/lib.rs` around lines 121 - 134, The Unix branch in
fake_ninja is still manually doing
metadata-to-permissions-to-set_mode-to-set_permissions, but the new
test_support::fs::set_mode helper already encapsulates that logic. Update the
fake_ninja setup code in test_support::lib::fake_ninja to call the helper
instead of re-implementing permission setting, and remove the now-unused
std::os::unix::fs::PermissionsExt import.

#### src/stdlib/which/lookup/tests.rs (1)

`100-121`: _🩺 Stability & Availability_ | _🟡 Minor_ | _⚡ Quick win_
**Restore directory permissions before asserting, not after.**
`test_fs::set_mode(blocked.as_std_path(), 0o700)` only runs once `.expect_err(...)` has already succeeded. If `search_workspace` unexpectedly returns `Ok`, the panic fires before the restore, leaving `blocked` at `0o000`. That can break `TempDir`'s recursive cleanup on drop and mask the real failure behind an unrelated permission error. Move the restore ahead of the unwrap so cleanup always happens.

##### 🔧 Proposed fix: restore mode before asserting

```diff
-    let err = search_workspace(&snapshot, "tool", false, &WorkspaceSkipList::default())
-        .expect_err("unreadable workspace entries should fail");
-
-    test_fs::set_mode(blocked.as_std_path(), 0o700).context("restore blocked")?;
+    let result = search_workspace(&snapshot, "tool", false, &WorkspaceSkipList::default());
+
+    test_fs::set_mode(blocked.as_std_path(), 0o700).context("restore blocked")?;
+
+    let err = result.expect_err("unreadable workspace entries should fail");
```

##### 🤖 Detailed instructions

Verify each finding against current code. Fix only still-valid issues, skip the
rest with a brief reason, keep changes minimal, and validate.

In `@src/stdlib/which/lookup/tests.rs` around lines 100 - 121, The test
`search_workspace_surfaces_unreadable_entries` restores `blocked` permissions
only after `expect_err`, so a failure path can leave the directory at `0o000`
and break cleanup. Update the test flow around `search_workspace` and
`test_fs::set_mode` so the permission restore happens before asserting on the
result, ensuring cleanup always runs even if `search_workspace` unexpectedly
succeeds.

#### .github/workflows/ci.yml (1)

`46-65`: _🚀 Performance & Scalability_ | _🟠 Major_ | _⚡ Quick win_
**Cache the staged Whitaker artefacts too, not just the installer binary.**
Shift the cache `path:` list to cover the staged lint-library directory and the generated `whitaker`/`whitaker-ls` wrappers, or the stable leg keeps rebuilding and re-staging Whitaker on every single run despite a cache hit on the installer binary.

```
In .github/workflows/ci.yml around lines 46-65, extend the cache `path:` list
to also include the directory where whitaker-installer stages its built
libraries and generates the `whitaker`/`whitaker-ls` wrapper scripts (consult
the whitaker-installer/Whitaker documentation, or run
`whitaker-installer --skip-wrapper` output locally, to determine the exact
staged-artefact and wrapper-script directories for the runner's platform,
since these are resolved via the crate's internal `dirs` module rather than
being fixed paths). Keep the cache key tied to WHITAKER_INSTALLER_VERSION
(and runner.os/runner.arch) so a version bump still invalidates stale caches.
Confirm a warm cache hit skips the rebuild/staging step entirely on the
stable matrix leg.
```

##### 🤖 Detailed instructions

Verify each finding against current code. Fix only still-valid issues, skip the
rest with a brief reason, keep changes minimal, and validate.

In @.github/workflows/ci.yml around lines 46 - 65, The cache in the CI workflow
only covers the whitaker-installer binary, so the stable job still rebuilds and
re-stages Whitaker artifacts on every run. Update the actions/cache
configuration in the Whitaker install step to also include the installer’s
staged library directory and the generated whitaker/whitaker-ls wrapper
locations, using the whitaker-installer path resolution/dirs behavior to choose
the correct platform-specific paths. Keep the cache key tied to runner.os,
runner.arch, and WHITAKER_INSTALLER_VERSION so version bumps invalidate stale
artifacts.

**Comment by leynos** [comment]
Recording a review-response decision: the suggestion to extend the Whitaker cache to the staged suite and wrappers is deliberately not adopted. Verified against the whitaker-installer source: the clone lands in `~/.local/share/whitaker` (`directories-next` `ProjectDirs::data_dir()`), prebuilt lint libraries stage at `~/.local/share/whitaker/lints/<toolchain>/<target>/lib` (ADR-001), and wrappers in `~/.local/bin`. However, installer 0.2.5 downloads prebuilt artefacts rather than building (nothing is "rebuilt every run"), the whole install step measures ~24 s against ~152 s for the lint itself, and `whitaker-installer` runs unconditionally, so caching those paths would not skip it. Caching the suite keyed by the installer version would also pin the deliberately rolling lint suite — the stale-lints failure mode removed from mxd in leynos/mxd#363. A suite cache becomes appropriate once suite pinning via `--ref` (leynos/whitaker#272) is adopted, keyed by the pinned ref.
